libdistopia/src/distopia.cpp

```diff
@@ -588,6 +588,7 @@ namespace distopia {
                 b_src = &b[0];
             }
             if (HWY_UNLIKELY(na * nb < nlanes)) {
+                out_sub = new T[nlanes * nlanes];
```

Comment:
I may be missing something but why do we need to dispatch to a scalar version if we are doing this expansion anyway? Why can we not just make slightly larger and `memcpy` the results out?

Reply:
the way we currently deal with partial lanes is to only advance the loading/saving pointer so that the final value is at the end of a vector.  e.g. for calc_bonds if we had 6 pairs coordinates and a vector width of 4, you'd calculated ABCD then CDEF.  So you've calculated and saved CD twice but that's not a big deal.
```
ABCDEF
xxxx..
..xxxx
```

it's a little different for distance array, where if you had a 4x3 problem size, and a vector width of 4 you couldn't just "rewind" and save backwards as you'd end up trampling previous row results.  To avoid writing a lot of code to cater for this, I'd opted for just reverting to scalar for small problem sizes since the speed is irrelevant.

Reply:
https://github.com/richardjgowers/distopia_2_the_highway_warrior/pull/25/files#diff-9517ca8b99329ab75ee01addf318fee26ef56e21f0299c909eaca9d8ccf2d78cL612 this line is where we're cautiously advancing the relevant pointer, which means we only need one loop over b.  You could instead have a full vector loop then a leftovers loop afterwards, which I think we previously have had.  This might be faster, but it's not been tried out yet.